src/hotspot/share/gc/g1/g1BufferListFreePool.hpp

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright (c) 2021, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#ifndef SHARE_GC_G1_G1BUFFERLISTFREEPOOL_HPP
+#define SHARE_GC_G1_G1BUFFERLISTFREEPOOL_HPP
+
+#include "gc/g1/g1CardSet.hpp"
+#include "gc/g1/g1SegmentedArray.hpp"
+#include "utilities/growableArray.hpp"
+
+// Statistics for a fixed set of buffer lists. Contains the number of buffers and memory
+// used for each. Note that statistics are typically not taken atomically so there
+// can be inconsistencies. The user must be prepared for them.
+class G1BufferListMemoryStats {
+public:
+
+  size_t _num_mem_sizes[G1CardSetConfiguration::num_mem_object_types()];
+  size_t _num_buffers[G1CardSetConfiguration::num_mem_object_types()];
+
+  // Returns all-zero statistics.
+  G1BufferListMemoryStats();
+
+  void add(G1BufferListMemoryStats const other) {
+    STATIC_ASSERT(ARRAY_SIZE(_num_buffers) == ARRAY_SIZE(_num_mem_sizes));
+    for (uint i = 0; i < ARRAY_SIZE(_num_mem_sizes); i++) {
+      _num_mem_sizes[i] += other._num_mem_sizes[i];
+      _num_buffers[i] += other._num_buffers[i];
+    }
+  }
+
+  void clear();
+
+  uint num_pools() const { return G1CardSetConfiguration::num_mem_object_types(); }
+};
+
+// A set of free lists holding memory buffers for use by G1CardSetAllocators.
+template<MEMFLAGS flag>
+class G1BufferListFreePool {
```

Comment:
I would prefer `G1SegmentedArrayFreePool` here, seeing the segmented array as the important thing (I understand that `G1SegmentedArrayBufferListFreePool` is too long).
I think there is already a CR to reconsider naming of all these classes, maybe add `G1SegmentedArrayBufferList` to the ones that need renaming; in particular I think right now that the "BufferList" of `G1SegmentedArrayBufferList` is kind of redundant.
Sorry for not speaking up earlier about this, but it only came to my mind right now.

It would be great for that renaming CR to start a thread in hotspot-gc-dev before actually doing the renames. Assuming that this renaming will be a close follow-up I do not have a particular opinion about the current name chosen, but maybe it is useful to actually start this discussion now to avoid renaming this class in particular again.

Reply:
Thanks Thomas.

> I would prefer `G1SegmentedArrayFreePool` here, seeing the segmented array as the important thing (I understand that `G1SegmentedArrayBufferListFreePool` is too long). 

Sure.

> I think there is already a CR to reconsider naming of all these classes, 

Do you mean https://bugs.openjdk.java.net/browse/JDK-8276299 or some other issue?

> maybe add `G1SegmentedArrayBufferList` to the ones that need renaming; 

If you mean 8276299 above, sure I will add this to the renaming list of 8276299.

> in particular I think right now that the "BufferList" of `G1SegmentedArrayBufferList` is kind of redundant. 

Not quite get your point, would you mind to clarify further?

> Sorry for not speaking up earlier about this, but it only came to my mind right now.

It's OK, it's never late to make necessary changes. :)

> It would be great for that renaming CR to start a thread in hotspot-gc-dev before actually doing the renames. 

Do you mean not send a pr but just send email to hotspot-gc-dev to discuss this idea of renaming?

> Assuming that this renaming will be a close follow-up I do not have a particular opinion about the current name chosen, but maybe it is useful to actually start this discussion now to avoid renaming this class in particular again.

If you mean 8276299 above, then I think these 2 (8276299 and this one) are not quite related to each other, 8276299 is to unify the naming of "buffer", "node" and "element" (especially in card set area currently), but this one is to factor out Factor out G1CardSetFreePool and enable evac failure to reuse the freeing memory logic in current G1CardSetFreePool.
Sure, I will start a discussion for 8276299 before actually send the pr.

Reply:
> Thanks Thomas.
> 
> > I would prefer `G1SegmentedArrayFreePool` here, seeing the segmented array as the important thing (I understand that `G1SegmentedArrayBufferListFreePool` is too long).
> 
> Sure.
> 
> > I think there is already a CR to reconsider naming of all these classes,
> 
> Do you mean https://bugs.openjdk.java.net/browse/JDK-8276299 or some other issue?

Yes.

> 
> > maybe add `G1SegmentedArrayBufferList` to the ones that need renaming;
> 
> If you mean 8276299 above, sure I will add this to the renaming list of 8276299.
> 
> > in particular I think right now that the "BufferList" of `G1SegmentedArrayBufferList` is kind of redundant.
> 
> Not quite get your point, would you mind to clarify further?

A segmented array is an array consisting of segments (as opposed to a single huge array of something); a list of buffers, where a "buffer" is nothing but an array (of some element type) is almost the same. One is a set of chunks of elements (to explicitly use other words), the other too.

> > It would be great for that renaming CR to start a thread in hotspot-gc-dev before actually doing the renames.
> 
> Do you mean not send a pr but just send email to hotspot-gc-dev to discuss this idea of renaming?

Not to discuss the idea of renaming, 8276299 is pretty clear about that, but discuss the actual mappings from old name to new name. Maybe this will wake up more people to the discussion than hiding it in a PR :) 

> 
> > Assuming that this renaming will be a close follow-up I do not have a particular opinion about the current name chosen, but maybe it is useful to actually start this discussion now to avoid renaming this class in particular again.
> 
> If you mean 8276299 above, then I think these 2 (8276299 and this one) are not quite related to each other, 8276299 is to unify the naming of "buffer", "node" and "element" (especially in card set area currently), but this one is to factor out Factor out G1CardSetFreePool and enable evac failure to reuse the freeing memory logic in current G1CardSetFreePool. Sure, I will start a discussion for 8276299 before actually send the pr.

That's true, but we are renaming something that also contains a "Buffer" in it and seemingly related.

So instead of renaming this `G1BufferListFreePool` again to something else in 8276299, maybe we could save the time and immediately change it to the final name now, then in 8276299 do the same for the other classes.

Then 8276299 will hopefully go through more smoothly too if we already decided on the new names beforehand as well.

Reply:
Thanks for clarification, I see your point, will initiate an discussion later. :)

Reply:
Discussion thread is here: https://mail.openjdk.java.net/pipermail/hotspot-gc-dev/2021-November/037626.html